enhancements/cli_manager.md

```diff
@@ -0,0 +1,166 @@
+`oc adm release extract --command oc` and `oc image extract` for each CLI image, and a CR for each CLI will provide the CLI-manager information about each CLI image.
+
+Regardless of the index scenario above, CLI-manager will be given the location of each CLI's artifacts via the CLI's CR.    
+
+Option 2 plus Option 3 is the option we as a team are leaning toward.  It requires that each CLI publish artifacts to [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+then create an image similar to these: [example images for oc cli-manager](https://github.com/sallyom/clustercli/tree/master/hack/cli-images)
+The CLI manager will gather information from each CLI's CR about each CLI image.  In disconnected environments this will be through a mirrored local 
+registry.  Extracting binaries from a CLI image will be similar to how we currently extract the installer and oc from a release payload.
+`oc cli-manager` will enable installing CLIs with known-compatible versions for a cluster.  CLIs will be managed through extending the oc commands we currently
+have that extract `oc` and `openshift-install` binaries from the release payload.  The logic for `oc image extract` will be re-used and extended for
+extracting all available openshift CLIs. As an example, the following command is possible today:
+
+```console
+$ oc image extract quay.io/sallyom/kn:latest --path /tmp/amd64/kn.tar.gz:$HOME/openshift-clis/compressed/
+$ tar xvf $HOME/openshift-clis/compressed/kn.tar.gz -C $HOME/openshift-clis/
+```
+
+As laid out in this proposal, each CLI will provide a separate image.
+Each CLI is responsible for creating a CR to hold metadata that the CLI-manager will need.
+The CLI-manager will access the CRs to allow a cluster user to install OpenShift tools that are known compatible with the user's cluster version.    
+The WIP [ClusterCLI API](https://github.com/openshift/api/pull/588)    
+
+To opt in, a CLI will create a ClusterCLI CR via it's operator.  A CLI's operator can be nothing more than a means to apply the CR, 
+or if the CLI already has an operator, it's current operator can add the CR.  
+The [ConsoleCLIDownload CRD](https://github.com/openshift/api/blob/master/console/v1/0000_10_consoleclidownload.crd.yaml) will be extended and reused.
+Currently, a user can run `oc get consoleclidownloads oc-cli-downloads -o yaml` or `oc get consoleclidownloads odo-cli-downloads -o yaml` 
+to get a download link for `odo` or `oc`.  `oc cli-manager` will take this a step further to handle installing binaries on the user's system, 
+in the $HOME directory, that are known compatibe with the current cluster's version.
+
+The CLI manager will:
+- List available CLI plugins - their latest version and/or the installed version
+- Install appropriate versions of a chosen CLI for a given cluster
+- Update installed version of a given CLI if/when updates become available
+- The above will be provided as a subcommand of oc.
```

Comment:
This is mixing the command and the index/mechanism. Would be nice to explicitly call out three roles:
1. index - is the central place where the information about binaries is being registered and read from
2. consumer - is responsible for reading data from index (1) and extracting binaries based on the provided information
3. producer - is responsible for providing metadata|description of the binary and writing it to index (1) and providing an image

Reply:
Each CLI will produce an image to deliver its artifacts.  The image registry (quay) will be the central place - currently those images will be packaging artifacts from https://mirror.openshift.com/pub/openshift-v4/clients/  but this mechanism only will care about the image, not where the artifacts in the image came from.

Reply:
We don't care how and where the binaries come from per se. Quay is also orthogonal. For us only the index/central repository matters which holds information about what's available and its metadata.